src/main/java/org/breedinginsight/services/geno/impl/GigwaGenoServiceImpl.java

```diff
@@ -0,0 +1,709 @@
+        ProgramsApi programsApi = brAPIEndpointProvider.get(genoBrAPIClient, ProgramsApi.class);
+        ApiResponse<BrAPIProgramListResponse> brAPIProgramListResponseApiResponse = programsApi.programsGet(new ProgramQueryParams().programDbId(program.getKey()));
+
+        return brAPIProgramListResponseApiResponse.getBody().getResult().getData().size() == 1;
+    }
+
+    private List<BrAPISample> fetchSamples(BrAPIClient genoBrAPIClient, Program program, List<BrAPIObservationUnit> observationUnits) throws ApiException, InstantiationException {
+        log.debug("fetching samples for OUs");
+        if(observationUnits.isEmpty()) {
+            log.debug("No OUs were supplied, returning");
+            return new ArrayList<>();
+        }
+
+        SamplesApi samplesApi = brAPIEndpointProvider.get(genoBrAPIClient, SamplesApi.class);
+
+        BrAPISampleSearchRequest sampleSearchRequest = new BrAPISampleSearchRequest();
+
+        sampleSearchRequest.setGermplasmDbIds(observationUnits.stream().map(ou -> program.getKey() + "§" + Utilities.removeProgramKeyAndUnknownAdditionalData(ou.getObservationUnitName(), program.getKey())).collect(Collectors.toList()));
+
+        return brAPIDAOUtil.search(samplesApi::searchSamplesPost, samplesApi::searchSamplesSearchResultsDbIdGet, sampleSearchRequest);
+    }
+
+    private List<BrAPIObservationUnit> fetchObservationUnits(BrAPIClient phenoBrAPIClient, BrAPIGermplasm germplasm) throws ApiException, InstantiationException {
+        ObservationUnitsApi observationUnitsApi = brAPIEndpointProvider.get(phenoBrAPIClient, ObservationUnitsApi.class);
+
+        BrAPIObservationUnitSearchRequest searchRequest = new BrAPIObservationUnitSearchRequest();
+        searchRequest.addGermplasmDbIdsItem(germplasm.getGermplasmDbId());
+
+        return brAPIDAOUtil.search(observationUnitsApi::searchObservationunitsPost, observationUnitsApi::searchObservationunitsSearchResultsDbIdGet, searchRequest);
+    }
+
+    private List<BrAPIObservationUnit> fetchObservationUnits(BrAPIClient phenoBrAPIClient, UUID experimentId) throws ApiException, DoesNotExistException, InstantiationException {
+        log.debug("fetching observationUnits for experiment: " + experimentId);
+        TrialsApi trialsApi = brAPIEndpointProvider.get(phenoBrAPIClient, TrialsApi.class);
+        ApiResponse<BrAPITrialListResponse> brAPITrialListResponseApiResponse = trialsApi.trialsGet(new TrialQueryParams().externalReferenceID(experimentId.toString())
+                                                                                                                          .externalReferenceSource(Utilities.generateReferenceSource(referenceSource, ExternalReferenceSource.TRIALS)));
+
+        if(brAPITrialListResponseApiResponse.getBody().getResult().getData() != null && brAPITrialListResponseApiResponse.getBody().getResult().getData().size() == 1) {
```

Comment:
This will fail if more than one trial is found.

Reply:
https://breedinginsight.atlassian.net/browse/BI-1710 fixes this problem, so now BreedBase will search for trials by xref